File: moegirl_synopsis.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
moegirl_synopsis.py
-------------------
从萌娘百科（https://mzh.moegirl.org.cn）抓取动漫条目的剧情简介。

关键点：
1. 萌娘百科的 "Moe Skin" 把真正的 wiki 内容放在
   `<template id="MOE_SKIN_TEMPLATE_BODYCONTENT">…</template>` 里，
   由客户端 Vue 应用渲染到 `<div id="app">`。抓页面源码时必须先把这个
   template 里的 HTML 取出来，再用它作为解析根。
2. 其 MediaWiki API（/api.php）默认会返回 "Unauthorized API call"，
   因此不走 API，直接抓页面 HTML 解析。
3. 页面小节标题的结构是：
       <div class="mw-heading mw-heading3">
           <h3 id="剧情简介"><span>…anchor…</span>剧情简介</h3>
       </div>
   heading 的 `id` 直接写中文（或 MediaWiki 百分号转义形式），
   外面还套了一层 `<div class="mw-heading">`，同级/越级判断时要连带
   这层 wrapper 一起比较。

用法：
    python moegirl_synopsis.py                    # 跑内置 3 个测试条目
    python moegirl_synopsis.py 杀手青春 葬送的芙莉莲 ...
"""

from __future__ import annotations

import re
import sys
import time
from typing import Optional

import requests
from bs4 import BeautifulSoup, Tag

BASE_URL = "https://mzh.moegirl.org.cn/"
USER_AGENT = (
    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/124.0 Safari/537.36"
)

# 按优先级排列的 "简介类" 小节标题名
CANDIDATE_SECTIONS = [
    "剧情简介",
    "故事简介",
    "剧情概要",
    "故事概要",
    "故事梗概",
    "内容简介",
    "作品简介",
    "剧情介绍",
    "简介",
]

TEMPLATE_ID = "MOE_SKIN_TEMPLATE_BODYCONTENT"

# 清洗时剔除的杂项
DROP_SELECTORS = [
    "table", "style", "script",
    ".reference", ".mw-editsection",
    ".navbox", ".infobox", ".toc",
    ".thumb", ".gallery", ".hatnote",
]


class MoegirlClient:
    def __init__(self, base_url: str = BASE_URL, timeout: int = 30) -> None:
        self.base_url = base_url
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update({
            "User-Agent": USER_AGENT,
            "Accept": "text/html,application/xhtml+xml",
            "Accept-Language": "zh-CN,zh;q=0.9",
        })

    # ------------------------------------------------------------------
    # 入口
    # ------------------------------------------------------------------
    def get_synopsis(self, title: str) -> Optional[str]:
        html = self._fetch_page(title)
        if html is None:
            return None

        body_root = self._extract_template_root(html)
        if body_root is None:
            return None

        synopsis = self._extract_from_section(body_root)
        if synopsis:
            return synopsis

        # 兜底：抓正文开头几段
        return self._extract_lead(body_root)

    # ------------------------------------------------------------------
    # 网络
    # ------------------------------------------------------------------
    def _fetch_page(self, title: str) -> Optional[str]:
        url = self.base_url + requests.utils.quote(title, safe="")
        resp = self.session.get(url, timeout=self.timeout)
        if resp.status_code == 404:
            return None
        resp.raise_for_status()
        return resp.content.decode("utf-8", errors="replace")

    # ------------------------------------------------------------------
    # HTML 解析
    # ------------------------------------------------------------------
    @staticmethod
    def _extract_template_root(html: str) -> Optional[Tag]:
        """从页面 HTML 里提取 MOE_SKIN template 内的 .mw-parser-output 根节点。

        做法：
        1) 用正则先抓出 <template id="MOE_SKIN_TEMPLATE_BODYCONTENT">…</template>
           里的 HTML 字符串（避免不同 parser 对 template 的处理差异）。
        2) 用 BeautifulSoup 再解析这段字符串，找 .mw-parser-output。
        3) 如果页面没有这个 template（老版皮肤或其他站），回退到整页 parse。
        """
        m = re.search(
            r'<template[^>]*id=["\']' + re.escape(TEMPLATE_ID) + r'["\'][^>]*>'
            r'(.*?)</template>',
            html, flags=re.DOTALL,
        )
        if m:
            inner = m.group(1)
            inner_soup = BeautifulSoup(inner, "html.parser")
            root = inner_soup.find(class_="mw-parser-output")
            return root or inner_soup

        # 回退：整页解析
        soup = BeautifulSoup(html, "html.parser")
        return soup.find(class_="mw-parser-output")

    def _extract_from_section(self, root: Tag) -> Optional[str]:
        heading = self._find_synopsis_heading(root)
        if heading is None:
            return None

        start_node = heading.find_parent("div", class_="mw-heading") or heading
        start_level = int(heading.name[1])

        parts: list[str] = []
        for sib in start_node.find_next_siblings():
            if self._is_heading_at_or_above(sib, start_level):
                break
            if sib.name in {"p", "ul", "ol", "blockquote"}:
                txt = sib.get_text("\n", strip=True)
                if txt:
                    parts.append(txt)
            elif sib.name == "div":
                # 某些 wiki 里用 <div> 排版的整段说明也应该收
                if any(cls in (sib.get("class") or [])
                       for cls in ("poem", "quote", "mw-collapsible")):
                    txt = sib.get_text("\n", strip=True)
                    if txt:
                        parts.append(txt)

        return self._tidy("\n\n".join(parts)) or None

    def _extract_lead(self, root: Tag) -> Optional[str]:
        self._strip_noise(root)
        paragraphs: list[str] = []
        for p in root.find_all("p", recursive=True):
            txt = p.get_text(" ", strip=True)
            if len(txt) < 10:
                continue
            paragraphs.append(txt)
            if len(paragraphs) >= 3:
                break
        return self._tidy("\n\n".join(paragraphs)) or None

    # ------------------------------------------------------------------
    # 工具
    # ------------------------------------------------------------------
    @staticmethod
    def _find_synopsis_heading(root: Tag) -> Optional[Tag]:
        headings = root.find_all(re.compile(r"^h[1-6]$"))
        # 1) 按候选关键词精确匹配
        for keyword in CANDIDATE_SECTIONS:
            for h in headings:
                if h.get("id") == keyword or h.get_text(strip=True) == keyword:
                    return h
        # 2) 模糊匹配
        for h in headings:
            text = h.get_text(strip=True)
            if re.search(r"(简介|概要|梗概)", text):
                return h
        return None

    @staticmethod
    def _is_heading_at_or_above(tag: Tag, level: int) -> bool:
        """判断 tag 是否是一个 <= level 的标题（或裹着 heading 的 mw-heading div）。"""
        if not isinstance(tag, Tag):
            return False
        if re.fullmatch(r"h[1-6]", tag.name or ""):
            return int(tag.name[1]) <= level
        if tag.name == "div" and "mw-heading" in (tag.get("class") or []):
            inner = tag.find(re.compile(r"^h[1-6]$"))
            if inner:
                return int(inner.name[1]) <= level
        return False

    @staticmethod
    def _strip_noise(root: Tag) -> None:
        for sel in DROP_SELECTORS:
            for tag in root.select(sel):
                tag.decompose()

    @staticmethod
    def _tidy(text: str) -> str:
        text = re.sub(r"\[\s*\d+\s*\]", "", text)          # 引用脚注 [1]
        text = re.sub(r"\[\s*编辑\s*\]", "", text)          # 编辑按钮
        text = re.sub(r"[ \t\u3000]+", " ", text)
        text = re.sub(r"\n{3,}", "\n\n", text)
        return text.strip()


# ----------------------------------------------------------------------
# CLI
# ----------------------------------------------------------------------
def main(argv: list[str]) -> int:
    titles = argv[1:] or ["杀手青春", "冰之城墙", "ATRI -My Dear Moments-"]
    client = MoegirlClient()

    hit = 0
    for idx, title in enumerate(titles, 1):
        print("=" * 72)
        print(f"[{idx}/{len(titles)}] 《{title}》")
        print("=" * 72)

        try:
            synopsis = client.get_synopsis(title)
        except requests.RequestException as exc:
            print(f"[网络错误] {exc}")
            synopsis = None

        if synopsis:
            hit += 1
            print(synopsis)
        else:
            print("[未找到简介]")

        print()
        time.sleep(0.4)          # 温和限速

    print(f"共 {len(titles)} 个条目，成功抓到简介 {hit} 个。")
    return 0 if hit == len(titles) else 1


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: moegirl_synopsis.mjs
#!/usr/bin/env node
/**
 * moegirl_synopsis.js
 * -------------------
 * 从萌娘百科（https://mzh.moegirl.org.cn）抓取动漫条目的剧情简介，
 * 支持可选地先从 bgm.tv 拿到「中文名 + 别名」作为候选标题。
 *
 * 关键点（踩坑记录）：
 * 1. 萌娘百科 "Moe Skin" 会把真正的 wiki 内容放在
 *        <template id="MOE_SKIN_TEMPLATE_BODYCONTENT">...</template>
 *    里，由客户端 Vue 渲染。直接从原始 HTML 解析正文时要先把这个
 *    template 的 innerHTML 取出来，再当作一份新的 HTML 文档解析。
 * 2. /api.php 默认会回 "Unauthorized API call"，所以直接抓页面 HTML。
 * 3. 小节标题结构（现代 MediaWiki）：
 *        <div class="mw-heading mw-heading3">
 *            <h3 id="剧情简介"><span id="..."></span>剧情简介</h3>
 *        </div>
 *    - id 在 h3 本身，不是 .mw-headline span
 *    - 外面套了一层 div.mw-heading，在"走兄弟节点找下一 heading"时要
 *      把这层 wrapper 也算进去
 * 4. 很多条目的中文名 / 别名和萌娘百科的页面名对不上，比如
 *    bgm 的《光之美少女》(subject 4243)，别名第一个是「两个人是光之美少女」，
 *    但萌娘百科里真正的那个页面叫《光之美少女(无印)》，
 *    直接搜"光之美少女"会落在「光之美少女系列」这个系列介绍页。
 *    本脚本的处理方式：
 *      a. 用 displayTitle 尾缀 `系列` / 页面内 `#系列介绍`/`#系列作品` /
 *         只有 `第X代` 分节而没有「简介」分节 这几条信号识别系列页；
 *      b. 在系列页里扫 <a title="..."> 的 title 属性 + 文字，
 *         只要文字命中 bgm 的任一别名，就把 title 当作真正的页面名
 *         再跳一次。
 *
 * 依赖：
 *     Node.js >= 18（用内置 fetch）
 *     cheerio（npm i cheerio）
 *
 * 用法：
 *     node moegirl_synopsis.js                 # 跑内置测试集
 *     node moegirl_synopsis.js 杀手青春        # 直接用萌娘百科标题
 *     node moegirl_synopsis.js 581284          # 纯数字当作 bgm subject id
 *     node moegirl_synopsis.js https://bgm.tv/subject/4243
 */

import * as cheerio from 'cheerio';

// ---------------------------------------------------------------------------
// 常量
// ---------------------------------------------------------------------------
const MOEGIRL_BASE = 'https://mzh.moegirl.org.cn/';
const BGM_BASE = 'https://bgm.tv/';

const USER_AGENT =
  'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 ' +
  '(KHTML, like Gecko) Chrome/124.0 Safari/537.36';

// 按优先级排序的简介类标题
const SYNOPSIS_KEYWORDS = [
  '剧情简介',
  '故事简介',
  '剧情概要',
  '故事概要',
  '故事梗概',
  '内容简介',
  '作品简介',
  '剧情介绍',
  '简介',
];

const TEMPLATE_ID = 'MOE_SKIN_TEMPLATE_BODYCONTENT';
const PAGE_DATA_ID = 'MOE_SKIN_PAGE_DATA';

// ---------------------------------------------------------------------------
// 网络
// ---------------------------------------------------------------------------
async function fetchHtml(url) {
  const res = await fetch(url, {
    headers: {
      'User-Agent': USER_AGENT,
      Accept: 'text/html,application/xhtml+xml',
      'Accept-Language': 'zh-CN,zh;q=0.9',
    },
    redirect: 'follow',
  });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error(`HTTP ${res.status} for ${url}`);
  return await res.text();
}

// ---------------------------------------------------------------------------
// 萌娘百科页面解析
// ---------------------------------------------------------------------------
/**
 * 把页面 HTML 拆成：真正渲染用的正文根 (.mw-parser-output) + displayTitle。
 */
function loadMoegirlPage(html) {
  const $full = cheerio.load(html);

  // displayTitle：优先取 MOE_SKIN_PAGE_DATA 里的 JSON
  let displayTitle = null;
  const pageDataEl = $full(`#${PAGE_DATA_ID}`);
  if (pageDataEl.length) {
    try {
      const raw = pageDataEl.text() || pageDataEl.html() || '{}';
      const data = JSON.parse(raw);
      displayTitle =
        data.displaytitle || data.wgPageName || data.wgTitle || data.title || null;
    } catch {
      /* 忽略 */
    }
  }
  if (!displayTitle) {
    const t = $full('title').text();
    if (t) displayTitle = t.replace(/\s*-\s*萌娘百科.*$/, '').trim() || null;
  }
  if (displayTitle) {
    displayTitle = displayTitle.replace(/<[^>]+>/g, '').trim();
  }

  // 真正的正文在 <template> 里
  const tplEl = $full(`#${TEMPLATE_ID}`);
  let $body;
  if (tplEl.length) {
    const inner = tplEl.html() || '';
    $body = cheerio.load(inner);
  } else {
    // 老版皮肤 / 其他站：直接回退到整页
    $body = $full;
  }

  let root = $body('.mw-parser-output').first();
  if (!root.length) root = $body.root();

  return { $: $body, root, displayTitle };
}

// ---------------------------------------------------------------------------
// 标题 / 系列页识别
// ---------------------------------------------------------------------------
function findSynopsisHeading($, root) {
  const headings = root.find('h1,h2,h3,h4,h5,h6').toArray();
  // 1) 精确匹配候选关键词（按优先级）
  for (const kw of SYNOPSIS_KEYWORDS) {
    for (const h of headings) {
      const $h = $(h);
      if ($h.attr('id') === kw || $h.text().trim() === kw) return $h;
    }
  }
  // 2) 模糊匹配
  for (const h of headings) {
    const text = $(h).text().trim();
    if (/(简介|概要|梗概)/.test(text)) return $(h);
  }
  return null;
}

function isSeriesPage($, root, displayTitle) {
  if (displayTitle && /系列\s*$/.test(displayTitle)) return true;
  if (root.find('#系列介绍, #系列作品').length) return true;

  const headings = root.find('h2,h3').toArray();
  const hasEra = headings.some((h) => {
    const $h = $(h);
    const id = $h.attr('id') || $h.text().trim();
    return /^第[一二三四五六七八九十百千]+代$/.test(id);
  });
  if (hasEra && !findSynopsisHeading($, root)) return true;

  return false;
}

/**
 * 走 heading 之后的兄弟节点，收集到下一个同级或更高级标题之前的段落文本。
 */
function extractSynopsis($, heading) {
  const $wrapper = heading.closest('div.mw-heading');
  const startNode = $wrapper.length ? $wrapper : heading;
  const startLevel = parseInt(heading.prop('tagName').substring(1), 10);

  const parts = [];
  let sib = startNode.next();

  while (sib.length) {
    const tag = (sib.prop('tagName') || '').toUpperCase();

    // 下一个同级或更高级 heading -> 停止
    if (tag === 'DIV' && sib.hasClass('mw-heading')) {
      const inner = sib.find('h1,h2,h3,h4,h5,h6').first();
      if (inner.length) {
        const lvl = parseInt(inner.prop('tagName').substring(1), 10);
        if (lvl <= startLevel) break;
      }
    } else if (/^H[1-6]$/.test(tag)) {
      const lvl = parseInt(tag.substring(1), 10);
      if (lvl <= startLevel) break;
    }

    if (['P', 'UL', 'OL', 'BLOCKQUOTE'].includes(tag)) {
      const t = sib.text().trim();
      if (t) parts.push(t);
    } else if (tag === 'DIV') {
      const cls = sib.attr('class') || '';
      if (/\b(poem|quote|mw-collapsible)\b/.test(cls)) {
        const t = sib.text().trim();
        if (t) parts.push(t);
      }
    }

    sib = sib.next();
  }

  return tidy(parts.join('\n\n'));
}

function tidy(text) {
  return text
    .replace(/\[\s*\d+\s*\]/g, '')
    .replace(/\[\s*编辑\s*\]/g, '')
    .replace(/[ \t\u3000]+/g, ' ')
    .replace(/\n{3,}/g, '\n\n')
    .trim();
}

/**
 * 系列页 -> 在页面里找 <a title="..."> 的文字命中别名的链接，
 * 把 title 属性当作真正的条目名返回。
 */
function findLinkByAlias($, root, aliases) {
  const anchors = root.find('a[title]').toArray();
  for (const alias of aliases) {
    if (!alias) continue;
    for (const a of anchors) {
      const $a = $(a);
      if ($a.text().trim() === alias) {
        const title = $a.attr('title');
        if (title) return title;
      }
    }
  }
  return null;
}

// ---------------------------------------------------------------------------
// bgm.tv 解析
// ---------------------------------------------------------------------------
async function fetchBgmInfo(subjectId) {
  const url = `${BGM_BASE}subject/${subjectId}`;
  const html = await fetchHtml(url);
  if (!html) return null;

  const $ = cheerio.load(html);
  const out = { cn: null, original: null, aliases: [] };

  const h1 = $('h1.nameSingle');
  if (h1.length) {
    const a = h1.find('a').first();
    if (a.length) out.original = a.text().trim();
  }

  const seen = new Set();
  $('#infobox li').each((_, li) => {
    const $li = $(li);
    const tip = $li.children('.tip').first();
    if (!tip.length) return;
    const key = tip.text().trim();

    // 从 .tip 之后开始取文本，遇到 ul/ol/br 就停（避免把嵌套的别名列表吞掉）
    let value = '';
    let n = tip[0].next;
    while (n) {
      if (n.type === 'tag' && ['ul', 'ol', 'br'].includes(n.name)) break;
      if (n.type === 'text') value += n.data;
      else if (n.type === 'tag') value += $(n).text();
      n = n.next;
    }
    value = value.trim();
    if (!value) return;

    if (key.startsWith('中文名')) {
      if (!out.cn) out.cn = value;
    } else if (key.startsWith('别名')) {
      if (!seen.has(value)) {
        seen.add(value);
        out.aliases.push(value);
      }
    }
  });

  return out;
}

// ---------------------------------------------------------------------------
// 主流程
// ---------------------------------------------------------------------------
async function tryPage(title) {
  const html = await fetchHtml(MOEGIRL_BASE + encodeURIComponent(title));
  if (!html) return { ok: false, reason: '404' };
  const page = loadMoegirlPage(html);
  return { ok: true, ...page };
}

async function resolveSynopsis(bgmInfo) {
  // 候选标题：中文名 > 每个别名
  const candidates = [];
  const push = (s) => {
    if (s && !candidates.includes(s)) candidates.push(s);
  };
  push(bgmInfo.cn);
  for (const a of bgmInfo.aliases) push(a);

  const allNames = candidates.slice(); // 用于系列页 link 匹配

  for (const candidate of candidates) {
    const page = await tryPage(candidate);
    if (!page.ok) continue;

    const { $, root, displayTitle } = page;
    if (!root || root.length === 0) continue;

    if (isSeriesPage($, root, displayTitle)) {
      const resolved = findLinkByAlias($, root, allNames);
      if (!resolved) continue;

      const deeper = await tryPage(resolved);
      if (!deeper.ok) continue;
      if (!deeper.root || deeper.root.length === 0) continue;
      if (isSeriesPage(deeper.$, deeper.root, deeper.displayTitle)) continue;

      const h = findSynopsisHeading(deeper.$, deeper.root);
      if (!h) continue;
      const syn = extractSynopsis(deeper.$, h);
      if (syn) {
        return {
          matchedTitle: resolved,
          via: `series(${candidate})->resolve`,
          synopsis: syn,
        };
      }
      continue;
    }

    const heading = findSynopsisHeading($, root);
    if (heading) {
      const syn = extractSynopsis($, heading);
      if (syn) {
        return {
          matchedTitle: displayTitle || candidate,
          via: 'direct',
          synopsis: syn,
        };
      }
    }
  }

  return null;
}

// ---------------------------------------------------------------------------
// CLI
// ---------------------------------------------------------------------------
const TEST_CASES = [
  { title: '杀手青春', bgmId: '581284' },
  { title: '冰之城墙' },
  { title: '亚托莉' },
  { title: '光之美少女', bgmId: '4243' }, // 系列页 + 别名解析的典型例子
];

function parseArg(arg) {
  const m = arg.match(/\/subject\/(\d+)/);
  if (m) return { bgmId: m[1] };
  if (/^\d+$/.test(arg)) return { bgmId: arg };
  return { title: arg };
}

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

async function runOne(target) {
  const label = target.title || target.bgmId || '?';
  console.log('='.repeat(72));
  console.log(`《${label}》${target.bgmId ? ` (bgm ${target.bgmId})` : ''}`);
  console.log('='.repeat(72));

  let bgmInfo = { cn: null, original: null, aliases: [] };
  if (target.bgmId) {
    try {
      bgmInfo = (await fetchBgmInfo(target.bgmId)) || bgmInfo;
      console.log(
        `[bgm] 中文名=${bgmInfo.cn ?? '(空)'}  原名=${bgmInfo.original ?? '(空)'}  别名=${JSON.stringify(bgmInfo.aliases)}`,
      );
    } catch (e) {
      console.log(`[bgm] 抓取失败: ${e.message}`);
    }
  }
  if (!bgmInfo.cn && target.title) bgmInfo.cn = target.title;

  let result = null;
  try {
    result = await resolveSynopsis(bgmInfo);
  } catch (e) {
    console.log(`[错误] ${e.message}`);
  }

  if (result) {
    console.log(`[匹配] ${result.matchedTitle}  (via ${result.via})`);
    console.log(`[字数] ${result.synopsis.length}`);
    console.log();
    console.log(result.synopsis);
  } else {
    console.log('[未找到简介]');
  }
  console.log();
}

async function main() {
  const args = process.argv.slice(2);
  const targets = args.length ? args.map(parseArg) : TEST_CASES;

  let hit = 0;
  for (const t of targets) {
    try {
      await runOne(t);
      hit += 1;
    } catch (e) {
      console.error(e);
    }
    await sleep(400); // 温和限速
  }
  console.log(`共 ${targets.length} 个，跑完 ${hit} 个。`);
}

// ESM entrypoint detect
const isMain =
  import.meta.url === `file://${process.argv[1]}` ||
  import.meta.url.endsWith(process.argv[1] || '');

if (isMain) {
  main().catch((e) => {
    console.error(e);
    process.exit(1);
  });
}

export {
  fetchBgmInfo,
  fetchHtml,
  loadMoegirlPage,
  findSynopsisHeading,
  isSeriesPage,
  extractSynopsis,
  findLinkByAlias,
  resolveSynopsis,
};